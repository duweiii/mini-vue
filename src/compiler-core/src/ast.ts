export const enum ENodeType {
  INTERPOLATION = 'interpolation',
  SIMPLE_EXPRESSION = 'simple_expression',
  ELEMENT = 'element',
  TEXT = 'text',
  ROOT = 'root',
  COMPOUND_EXPRESSION = 'compound_expression'
}

export const enum EElementStatus {
  START,
  END
}

export function createVNodeCall(tag, props, children){
    return {
      type: ENodeType.ELEMENT,
      tag,
      props,
      children,
    }
}