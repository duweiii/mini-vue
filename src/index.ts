export * from './runtime-dom/index'
export * from './reactiviy/index'