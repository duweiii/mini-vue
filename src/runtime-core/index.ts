export { h } from "./h";
export { renderSlots } from './helpers/renderSlots'
export { createTextVNode } from './createVNode'
export { getCurrentInstance } from './component'
export { provide, inject } from './apiInject'
export { createRenderer } from './render'
export { nextTick } from './schedulers'