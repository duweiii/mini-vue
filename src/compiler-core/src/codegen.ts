import { isString } from "../../shared/index";
import { ENodeType } from "./ast";
import { CREATE_ELEMENT_VNODE, helperMap, TO_DISPLAY_STRING } from "./runtimeHelpers";

export function generate(node){
  const context = createCodegenContext()
  const { push } = context;
  const functionName = 'render';
  const args = ['_ctx', '_cache'];
  const signature = args.join(',');
  
  if( node.helpers.length > 0 ){
    genFunctionPreamble(node, context)
  }

  push('return function ')
  push(`${functionName}(${signature}){return `)
  genNode(node.codegenNode, context)
  push("}")

  return {
    code: context.code
  }
}

function genFunctionPreamble(node, context){
  const { push } = context;
  const VueBinging = 'Vue';
  const aliasHelpers = (s) => `${helperMap[s]}:_${helperMap[s]}`;
  const helpers = node.helpers;
  const declaration = `const { ${helpers.map(aliasHelpers)} } = ${VueBinging}`
  push(declaration)
  push('\n')
}

function createCodegenContext(){
  const context = {
    code: '',
    push(source){
      context.code += source
    },
    helper(key){
      return `_${helperMap[key]}`;
    }
  }
  return context;
}

function genNode(node, context){
  switch( node.type ){
    case ENodeType.TEXT:
      genText(node, context);
      break;
    case ENodeType.INTERPOLATION:
      genInterpolation(node, context)
      break;
    case ENodeType.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break;
    case ENodeType.ELEMENT:
      genElement(node, context);
      break;
    case ENodeType.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
}

function genCompoundExpression(node, context){
  const { children } = node;
  const { push } = context;
  for (let i = 0; i < children.length; i++) {
    let child = children[i];
    if( isString(child) ){
      push(child);
    }else{
      genNode(child, context)
    }
  }
}

function genElement(node, context){
  const { helper, push } = context;
  const { tag, props, children } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  genNodeList( genNullable([tag, props, children]), context );
  push(")")
}

function genNullable(args){
  return args.map(arg => arg || 'null');
}

function genNodeList(nodes, context){
  const { push } = context;
  for (let index = 0; index < nodes.length; index++) {
    const node = nodes[index];
    if( isString(node) ){
      push(node);
    }else{
      genNode(node, context);
    }
    if( index < nodes.length - 1 ){
      push(', ')
    }
  }
}

function genText(node, context){
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node, context){
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode( node.content, context )
  push(')')
}

function genExpression(node, context){
  const { push } = context;
  push( node.content )
}