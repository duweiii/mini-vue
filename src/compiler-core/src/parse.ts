import { EElementStatus, ENodeType } from "./ast";

export function baseParse(content){
  let context = createContext(content);
  return createRoot( parseChildren(context, []) )
}
function createRoot(children){
  return {
    type: ENodeType.ROOT,
    children
  }
}
function createContext(content){
  return {
    source: content
  }
}
function parseChildren(context, ancestor){
  let nodes: any = [];
  let node;
  while( !isEnd(context, ancestor) ){
    if( context.source.startsWith("{{")){
      node = parseInterpolation(context)
    }else if( context.source[0] === '<' ){
      if( /[a-z]/i.test( context.source[1] ) ){
        node = parseElement( context , ancestor)
      }
    }
  
    if( !node ){
      node = parseText(context)
    }
  
    nodes.push(node)
  }
  return nodes;
}
function isEnd(context, ancestor){
  /**
   * 结束的条件
   * 1. context.source 为空
   * 2. 遇到结束标签了
   */
  let s = context.source;
  if( s.startsWith("</") ){
    for( let i = ancestor.length - 1; i >= 0 ; i-- ){
      let tag = ancestor[i];
      if( startsWithCloseTagAndIsSameTag(context, tag) ){
        return true;
      }
    }
  }

  return !s;
}
function startsWithCloseTagAndIsSameTag(context, tag){
  return context.source.startsWith("</") && tag.toLowerCase() === context.source.slice(2, 2 + tag.length).toLowerCase();
}
function parseElement(context, ancestor){
  let element: any = parseTag( context, EElementStatus.START)
  ancestor.push( element.tag )
  element.children = parseChildren(context, ancestor)
  ancestor.pop()
  if( startsWithCloseTagAndIsSameTag(context, element.tag) && !loseEndTag(context, ancestor, element.tag) ){
    // 还有 <div> <div> </div> 这种情况需要处理
    // 这时，ancestor栈中的div数量应该比context.source中少一个
    // 所以当 ancestor>div's count === context.source中div的数量-1时，标签才是正常的

    /**
     * 上面说的情况现在的判断也可以捕获到缺少标签。
     * <div><span></div>
     * 这种情况，因为span跟后面的div不匹配，所以肯定不会命中这个if，所以走到else中
     * <div><div></div>
     * 这种情况，中间的div调用startsWithCloseTagAndIsSameTag这个判断，结果为true，
     * 会把后面的父元素的结束标签删掉，
     * 然后父元素在最后需要处理结束标签时，因为 context.source 已经为空字符串了，
     * 所以肯定不会等于element.tag，也会进入else的逻辑报错。
     * 
     * 所以最后的问题就是，进入错误逻辑的本该是子div元素，现在是父div元素。错误定位不准确。
     * 
     * 添加判断 !loseEndTag(context, ancestor, element.tag)
     */
    parseTag( context, EElementStatus.END)
  }else{
    throw new Error(`缺少结束标签${element.tag}`);
  }
  return element;
}
function loseEndTag(context, ancestor, tag){
  const duplicateString = context.source;
  const endTagCountInSourceString = duplicateString.split(`</${tag}>`).length - 1;
  let tagCountInAncestor = 0;
  ancestor.forEach(item => {
    item === tag && tagCountInAncestor++;
  })
  return tagCountInAncestor + 1 !== endTagCountInSourceString;
}
function parseTag(context, type: EElementStatus){
  let match: any = /^<\/?([a-z]*)/i.exec( context.source );
  let tag = match[1];
  advanced(context, match[0].length)
  advanced(context, 1)
  if(type === EElementStatus.END) return ;
  return {
    type: ENodeType.ELEMENT,
    tag,
  }
}
function parseText(context){
  let endIndex = context.source.length;
  let endSymbol = ["<","{{"]

  endSymbol.forEach((item) => {
    let index = context.source.indexOf(item);
    if( index !== -1 && index < endIndex){
      endIndex = index;
    }
  })
  let content = parseTextData(context, endIndex)
  return {
    type: ENodeType.TEXT,
    content,
  }
}
function parseInterpolation(context){
  let openDelimiter = "{{"
  let closeDelimiter = "}}"
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
  advanced(context, openDelimiter.length)
  let rawLength = closeIndex - openDelimiter.length;
  let rawContent = parseTextData(context, rawLength)
  let content = rawContent.trim()
  advanced(context, closeDelimiter.length)
  return {
    type: ENodeType.INTERPOLATION,
    content: {
      type: ENodeType.SIMPLE_EXPRESSION,
      content,
    }
  }
}
function parseTextData(context, length){
  let content = context.source.slice(0, length);
  advanced(context, length)
  return content;
}
function advanced(context, leng){
  context.source = context.source.slice(leng)
}