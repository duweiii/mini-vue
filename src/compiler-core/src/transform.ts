import { ENodeType } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root, options = {}){
  const context = createTransformContext(root, options)
  traverseNode( root, context );
  createCodegenNode( root );
  createHelpers(root, context);
}
function createHelpers(root, context){
  root.helpers = [...context.helpers.keys()];
}
function createCodegenNode( root ){
  root.codegenNode = root.children[0]
}
function createTransformContext(root, options){
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],    
    helpers: new Map(),
    helper(key){
      context.helpers.set(key, 1)
    }
  }
  return context;
}

function traverseNode(node, context){

  const { nodeTransforms } = context;
  let exitFns: any = [];

  for( let i = 0; i < nodeTransforms.length; i++ ){
    const transform = nodeTransforms[i];
    const cb = transform(node, context)
    cb && exitFns.push(cb);
  }

  switch(node.type){
    case ENodeType.ROOT:
    case ENodeType.ELEMENT:
      traverseChildren( node, context );
      break;
    case ENodeType.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
    default: 
      break;
  }

  let i = exitFns.length;
  while(i--){
    exitFns[i] && exitFns[i]();
  }
}

function traverseChildren( node, context ){
  const { children } = node;
  for( let i = 0; i < children.length ; i++ ){
    const node = children[i]
    traverseNode( node, context)
  }
}