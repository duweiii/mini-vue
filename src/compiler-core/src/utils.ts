import { ENodeType } from "./ast";

export function isText(node){
  return node.type === ENodeType.TEXT || node.type === ENodeType.INTERPOLATION;
}
