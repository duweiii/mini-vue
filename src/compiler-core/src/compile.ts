import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformChildren } from "./transforms/transformChildren";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

export function baseCompile(template: string){
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [ transformExpression, transformChildren, transformElement, transformText]
  })
  return generate(ast);
}