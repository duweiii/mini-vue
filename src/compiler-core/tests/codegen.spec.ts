import { generate } from "../src/codegen";
import { baseParse } from "../src/parse"
import { transform } from "../src/transform";

describe('code generate', () => {
  it('happy path', () => {
    const ast = baseParse('hi');
    transform(ast)
    const { code } = generate( ast );
    expect( code ).toMatchSnapshot();
  })
})