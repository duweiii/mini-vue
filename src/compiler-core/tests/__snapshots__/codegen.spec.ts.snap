// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generate happy path: interpolation node 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;

exports[`code generate happy path: text node 1`] = `"return function render(_ctx,_cache){return 'hi'}"`;

exports[`code generate happy path: unit 1`] = `
"const { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`code generate happy pathï¼šelement node 1`] = `
"const { createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, null)}"
`;
